src/osd/PeeringState.cc

```diff
@@ -525,6 +525,18 @@ bool PeeringState::should_restart_peering(
 	       << dendl;
     return true;
   }
+  if (osdmap->get_pg_size(info.pgid.pgid) >
```

Comment:
I don't think this works in general.  It's not an actual interval change as far as the clients are concerned, so clients won't resend ops.  Because different osds might have different past_intervals, it isn't even necessarily the case that all osds with this pg will consider it an interval change, so some osds may or may not even respond to peering messages resulting in a broken pg and/or crashed osds.

When osd.2 in your example comes back up, does it send an MOSDPGNotify to the primary?  I think the right way to make this work right is for

boost::statechart::result PeeringState::Active::react(const MNotifyRec& notevt)

to check whether the newly booted node could be validly chosen for the acting set and request a new temp mapping.  The new temp mapping would trigger a real interval change.  I'll note, however, that that's a significantly larger refactor and if osd.2 is down long enough it would need backfill anyway.

Reply:
> It's not an actual interval change as far as the clients are concerned, so clients won't resend ops. ... so some osds may or may not even respond to peering messages resulting in a broken pg and/or crashed osds.

No, it does not form a new interval change at this moment. But as primary restarting peering process, it will redo choose_acting, generate a new pg_temp  containing the stray peer that has come back and hence trigger a real interval change finally. Also there are some other exceptions need to be carefully handled.

I agree that this isn't a very ideally fix but we have locally verified that it (together with another small fix) turned out to work pretty well.

Anyway, thanks for your inputs.  Yan and I will try to figure out a better fix soon.

Reply:
It's not an actual interval change as far as the clients are concerned, so clients won't resend ops. ... so some osds may or may not even respond to peering messages resulting in a broken pg and/or crashed osds.

No, it does not form a new interval change at this moment. But as primary restarting peering process, it will redo choose_acting, generate a new pg_temp containing the stray peer that has come back and hence trigger a real interval change finally. Also there are some other exceptions need to be carefully handled.

I agree that this isn't a very ideal fix but we have locally verified that it (together with another small fix) turned out to work pretty well.

Anyway, thanks for your inputs. Yan and I will try to figure out a better fix soon.